flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/assigners/MySqlSnapshotSplitAssigner.java

```diff
@@ -169,13 +170,42 @@ private void captureNewlyAddedTables() {
             // check whether we got newly added tables
             try (JdbcConnection jdbc = openJdbcConnection(sourceConfig)) {
                 final List<TableId> newlyAddedTables = discoverCapturedTables(jdbc, sourceConfig);
+                List<TableId> tablesToRemove = new LinkedList<>(alreadyProcessedTables);
+                tablesToRemove.addAll(remainingTables);
+                tablesToRemove.removeAll(newlyAddedTables);
+
                 newlyAddedTables.removeAll(alreadyProcessedTables);
                 newlyAddedTables.removeAll(remainingTables);
+
+                // case 1: there are old tables to remove from state
+                if (!tablesToRemove.isEmpty()) {
+                    // remove unassigned tables/splits if it does not satisfy new table filter
+                    remainingTables.removeAll(tablesToRemove);
+                    List<String> splitsToRemove = new LinkedList<>();
+                    for (Map.Entry<String, MySqlSnapshotSplit> splitEntry :
+                            assignedSplits.entrySet()) {
+                        if (tablesToRemove.contains(splitEntry.getValue().getTableId())) {
+                            splitsToRemove.add(splitEntry.getKey());
+                        }
+                    }
+
+                    splitsToRemove.forEach(assignedSplits.keySet()::remove);
+                    splitsToRemove.forEach(splitFinishedOffsets.keySet()::remove);
+                    remainingSplits.removeIf(split -> tablesToRemove.contains(split.getTableId()));
+                    alreadyProcessedTables.removeIf(tableId -> tablesToRemove.contains(tableId));
+                }
+
+                // case 2: there are new tables to add
                 if (!newlyAddedTables.isEmpty()) {
                     // if job is still in snapshot reading phase, directly add all newly added
                     // tables
                     LOG.info("Found newly added tables, start capture newly added tables process");
+
+                    // add new tables
                     remainingTables.addAll(newlyAddedTables);
+
+                    // only suspend the assigner if there's new tables added and assigner has
```

Comment:
only suspend the assigner if the assigner in finished status

Reply:
Comment updated.